test/src/unit-cppapi-array.cc

```diff
@@ -2088,3 +2088,50 @@ TEST_CASE(
   vfs.remove_dir(get_commit_dir(array_read.uri()));
   vfs.remove_dir(array_read.uri() + "/__schema");
 }
+
+TEST_CASE(
+    "C++ API: Close array with running query", "[cppapi][close-before-read]") {
+  const std::string array_name = "cpp_unit_array";
+  Context ctx;
+  VFS vfs(ctx);
+
+  if (vfs.is_dir(array_name))
+    vfs.remove_dir(array_name);
+
+  // Create
+  Domain domain(ctx);
+  domain.add_dimension(Dimension::create<int>(ctx, "rows", {{0, 3}}, 4))
+      .add_dimension(Dimension::create<int>(ctx, "cols", {{0, 3}}, 4));
+  ArraySchema schema(ctx, TILEDB_DENSE);
+  schema.set_domain(domain).set_order({{TILEDB_ROW_MAJOR, TILEDB_ROW_MAJOR}});
+  schema.add_attribute(Attribute::create<int>(ctx, "a"));
+  Array::create(array_name, schema);
+
+  // Write
+  std::vector<int> data_w = {
+      1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16};
+  Array array_w(ctx, array_name, TILEDB_WRITE);
+  Query query_w(ctx, array_w);
+  query_w.set_subarray(Subarray(ctx, array_w).set_subarray({0, 3, 0, 3}))
+      .set_layout(TILEDB_ROW_MAJOR)
+      .set_data_buffer("a", data_w);
+  query_w.submit();
+  array_w.close();
+
+  // Open for read.
+  Array array(ctx, array_name, TILEDB_READ);
+  std::vector<int> subarray_read = {0, 3, 0, 3};
+  std::vector<int> a_read(16);
+
+  Query query(ctx, array);
+  query.set_subarray(subarray_read)
+      .set_layout(TILEDB_ROW_MAJOR)
+      .set_data_buffer("a", a_read);
+  query.submit_async();
+  array.close();
+
+  while (query.query_status() != Query::Status::COMPLETE) {
+  }
```

Comment:
I'd add an iteration counter to this loop so that we can assert that the query was actually running after the `array.close()` call returns. Otherwise its possible that the query was just so fast it was done before that close call executed.

```suggestion
  uint64_t iterations = 0;
  while (query.query_status() != Query::Status::COMPLETE) {
    iterations++;
  }
  REQUIRE(iterations > 0);
```